jax/_src/lax/lax.py

```diff
@@ -1088,19 +1091,29 @@ def argmax(operand: Array, axis: int,
   return argmax_p.bind(operand, axes=(axis,),
                        index_dtype=dtypes.canonicalize_dtype(index_dtype))
 
-def reduce(operand: Array, init_value: Array, computation: Callable,
-           dimensions: Sequence[int]) -> Array:
+def reduce(operands: Array, init_values: Array, computation: Callable,
+           dimensions: Sequence[int], **reduce_kwargs: Any) -> Array:
   """Wraps XLA's `Reduce
   <https://www.tensorflow.org/xla/operation_semantics#reduce>`_
   operator.
   """
-  monoid_reducer = _get_monoid_reducer(computation, init_value)
+  flat_operands, operand_tree = tree_util.tree_flatten(operands)
+  flat_init_values, init_value_tree = tree_util.tree_flatten(init_values)
+  assert operand_tree == init_value_tree, (
```

Comment:
I'd probably raise a TypeError or ValueError explicitly rather than using assert. I'd also include details of the structure to help the user. Same everywhere.

Reply:
Done.